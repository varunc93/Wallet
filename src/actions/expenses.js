import uuid from 'uuid';

//Action Generators

//Expenses

export const addExpense = ({description = '', text = '', amount = 0, createdAt = 0} = {}) => ({
    type: "ADD_EXPENSE",
    expense: {
        id: uuid(),
        description,
        text,
        amount,
        createdAt
    }
});

export const removeExpense = ({id} = {}) => ({
    type: "REMOVE_EXPENSE",
    id
});

export const editExpense = (id, updates) => ({   //another way of passing id
    type: "EDIT_EXPENSE",
    id,
    updates
});