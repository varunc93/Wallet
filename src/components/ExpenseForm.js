import 'react-dates/initialize';
import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';

export default class ExpenseForm extends React.Component {
    constructor(props) { //This is done to get access to props
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : "",
            amount: props.expense ? (props.expense.amount/100).toString() : "",
            text: props.expense ? props.expense.text : "",
            createdAt: props.expense ? moment(props.expense.createdAt): moment(),
            focused: false,
            error:""
        }
    };

    handleDescription = (event) => {
        const description = event.target.value;
        this.setState(() => ({description}));
    };

    //OR
    // handleDescription = (event) => {
    //     event.persist;
    //     this.setState(() => ({description: event.target.value}));
    // }

    handleAmount = (event) => {
        const amount = event.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) //We use regular expressions to check validation. ^-ensures the user starts with a number
        //\d-ensures it's a number, *- to allow multiple digits, ()-allows optional parameters,
        //?-ensures the parameter is only allowed once, \.-allows one decimal digit, {0,2}-allows upto 2 characters
        //$ ensures we end the input with the regular expression mentioned
        //match() takes in regular expression, /<place regular expression here>/
        // !amount allows user to clear the data in the amount box
        this.setState(() => ({amount}));
    };

    handleText = (event) => {
        const text = event.target.value;
        this.setState(() => ({text}));
    };

    onDateChange = (createdAt) => {
        if (createdAt) {
          this.setState(() => ({ createdAt }));
        }
    };

    onFocusChange = ({focused}) => {
        this.setState(() => ({focused}));
    };

    onSubmit = (event) => {
        event.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState((error) => true);
            if(!this.state.description){
               this.setState(() => ({error: "Error: Please enter description"}))
            }
            else{
                this.setState(() => ({error: "Error: Please enter amount"}))
            }
        }
        else {
            this.setState(() => ({error: ""}))
            this.props.onSubmit({ //Props to be passed to Add and Edit Expense Pages
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100, //Since amount is stored as string
                text: this.state.text,
                createdAt: this.state.createdAt.valueOf()
            });
        }
    };

    //Button is placed inside a div to prevent it from being a direct descendant.
    render () {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input className="text-input" type="text" placeholder="Description" autoFocus value={this.state.description} onChange={this.handleDescription}/>
                <input className="text-input" type="text" placeholder="Amount" value= {this.state.amount} onChange={this.handleAmount}/>
                <textarea className="textarea" placeholder="Add a note..." value={this.state.text} onChange={this.handleText}></textarea>
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.focused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <div>
                    <button className="button">Save Expense</button>
                </div>
            </form>
        )
    };

}
