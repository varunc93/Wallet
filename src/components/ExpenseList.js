import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem'
import getVisibleExpenses from '../selectors/expenses'

export const ExpenseList = (props) => ( //Export required for testing to isolate ExpenseList and use custom expenses
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
            {
                props.expenses.length === 0 ? (
                    <div className="list-item list-item--message">
                        <span>There are no expenses to show!</span>
                    </div>
                ):
                (
                    props.expenses.map((expense) => {
                        return <ExpenseListItem key={expense.id} {...expense} />;
                }))
            }
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    };
}

//Same as
// const mapStateToProps = (state) => ({
//     expenses: state.expenses,
//     filters: state.filters
// });


export default connect(mapStateToProps)(ExpenseList); //Connect returns a function