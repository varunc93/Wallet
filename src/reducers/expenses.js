//Expense Reducer

const expenseReducerDefault = [];

const expenseReducer = (state = expenseReducerDefault, action) => {
    switch(action.type){
        case "ADD_EXPENSE":
            return [...state, action.expense];                //Using spread operator does not change the existing array.
        case "REMOVE_EXPENSE":
            return state.filter(({id}) =>  id !== action.id);
        case "EDIT_EXPENSE":
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                }
            });
        case 'SET_EXPENSES':
            return action.expenses;
        default: return state;
    }
};

export default expenseReducer;