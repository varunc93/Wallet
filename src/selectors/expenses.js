import moment from 'moment';
//Selectors

const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {  //de-structure filters
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
        return textMatch && startDateMatch && endDateMatch;
    }).sort((a,b) => {
        if(sortBy.toLowerCase() === "date")
            return a.createdAt < b.createdAt ? 1 : -1;  //Print the most recent expense first, if a is created first, then print a last.
        else if(sortBy.toLowerCase() === "amount")
            return a.amount - b.amount < 0 ? 1 : -1; //Print higher amount first
        else
            return 0;
    });
};

export default getVisibleExpenses;