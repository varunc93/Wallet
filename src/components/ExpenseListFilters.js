import React from 'react';
import {connect} from 'react-redux';
import {DateRangePicker} from 'react-dates';
import {setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate} from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    state = {
        focused: null
    };

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (focused) => { //focused is the new value
        this.setState(() => ({focused})); //Same as this.setState(() => ({focused: focused}));
    };

    onTextChange = (event) => {
        this.props.setTextFilter(event.target.value);
    };

    onSortChange = (event) => {
        if(event.target.value === "date"){
            this.props.sortByDate();
        }
        else if(event.target.value === "amount"){
            this.props.sortByAmount();
        }
    };

    render() {
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onTextChange}/>
                <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.focused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    startDateId="start"
                    endDateId="end"
                />
            </div>
        )
    }
};

//Basic Template
// const mapStateToProps = (state) => {
//     return {

//     }
// }

const mapStateToProps = (state) => { //Access to whole and we need to filter out things we want and don't want and pass it through connect.
    return {
        filters: state.filters
    }
};

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount())
});


export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);