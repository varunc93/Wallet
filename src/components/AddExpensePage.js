import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../actions/expenses';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        // props.dispatch(addExpense(expense));
        this.props.addExpense(expense);
        this.props.history.push("/"); //push takes in a path to reroute to a stored path
    };

    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit = {this.onSubmit} //Props received from Expense Form
                />
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({addExpense: (expense) => dispatch(addExpense(expense))}) //To make it easier for testing purposes

export default connect(undefined, mapDispatchToProps)(AddExpensePage);