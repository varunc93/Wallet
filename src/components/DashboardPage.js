import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpenseSummary';

const DashboardPage = () => (
    <div>
        <ExpenseSummary />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default DashboardPage;