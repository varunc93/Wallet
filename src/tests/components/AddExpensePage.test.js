import React from 'react';
import {shallow} from 'enzyme';
import {AddExpensePage} from '../../components/AddExpensePage';
import expenses from '../fixtures/expenses';
import toJSON from 'enzyme-to-json';

let addExpense, history, wrapper;

beforeEach(() => {
    addExpense = jest.fn();
    history = {push: jest.fn()};
    wrapper = shallow(<AddExpensePage  addExpense={addExpense} history={history}/>);
});

test("Render AddExpense Correctly", () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test("Handle AddExpense onSubmit", () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2]);
    expect(history.push).toHaveBeenLastCalledWith("/");
    expect(addExpense).toHaveBeenLastCalledWith(expenses[2]);
    expect(toJSON(wrapper)).toMatchSnapshot();
});