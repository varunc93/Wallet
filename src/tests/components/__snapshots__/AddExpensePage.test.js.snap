// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handle AddExpense onSubmit 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Add Expense
    </h1>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;

exports[`Render AddExpense Correctly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Add Expense
    </h1>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
